samples/mixed/provider/src/provider_impl.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+use log::{debug, info, warn};
+use proto::consumer::{consumer_client::ConsumerClient, RespondRequest};
+use proto::provider::{
+    provider_server::Provider, GetRequest, GetResponse, InvokeRequest, InvokeResponse, SetRequest,
+    SetResponse, SubscribeRequest, SubscribeResponse, UnsubscribeRequest, UnsubscribeResponse,
+};
+use std::collections::{HashMap, HashSet};
+use std::str::FromStr;
+use std::sync::{Arc, Mutex, MutexGuard};
+use tonic::{Request, Response, Status};
+
+use crate::vehicle::Vehicle;
+
+pub type SubscriptionMap = HashMap<String, HashSet<String>>;
+
+/// The ids for commands.
+const ACTIVATE_AIR_CONDITIOING: &str =
+    "dtmi:org:eclipse:sdv:vehicle:cabin:hvac:activate_air_conditioning;1";
+const SEND_NOTIFICATION: &str = "dtmi:org:eclipse:sdv:vehicle:cabin:hvac:send_notification;1";
+const SET_UI_MESSAGE: &str = "dtmi:org:eclipse:sdv:vehicle:cabin:hvac:set_ui_message;1";
+
+#[derive(Debug, Default)]
+pub struct ProviderImpl {
+    pub subscription_map: Arc<Mutex<SubscriptionMap>>,
+    pub vehicle: Arc<Mutex<Vehicle>>,
+}
+
+impl ProviderImpl {
+    fn activate_air_conditioning(
+        vehicle: Arc<Mutex<Vehicle>>,
+        payload: &str,
+    ) -> Result<(), String> {
+        let value: bool = match FromStr::from_str(payload) {
+            Ok(v) => v,
+            Err(error) => return Err(format!("{error:?}")),
+        };
+
+        let mut lock: MutexGuard<Vehicle> = vehicle.lock().unwrap();
+
+        lock.is_air_conditioning_active = value;
+
+        Ok(())
+    }
+
+    fn send_notification(payload: &str) {
+        info!("Notification: '{payload}'");
+    }
+
+    fn set_ui_message(vehicle: Arc<Mutex<Vehicle>>, payload: &str) {
+        let mut lock: MutexGuard<Vehicle> = vehicle.lock().unwrap();
+
+        lock.ui_message = String::from(payload);
+    }
+}
+
+#[tonic::async_trait]
+impl Provider for ProviderImpl {
+    /// Subscribe implementation.
+    ///
+    /// # Arguments
+    /// * `request` - Subscribe request.
+    async fn subscribe(
+        &self,
+        request: Request<SubscribeRequest>,
+    ) -> Result<Response<SubscribeResponse>, Status> {
+        let request_inner = request.into_inner();
+        let entity_id: String = request_inner.entity_id.clone();
+        let consumer_uri: String = request_inner.consumer_uri;
+
+        info!(
+            "Received a subscribe request from consumer URI {} for entity id {}",
+            &consumer_uri, &entity_id
+        );
+
+        let mut lock: MutexGuard<HashMap<String, HashSet<String>>> =
```

Comment:
Here and probably other places where Locking mechanism is created on stack:

1. I don't think this locks the `self.subscription_map` because every async call to this `subscribe` function creates a new instance of the locking object on stack and locks it successfully, because it is unique per stack frame. To have concurrency, the locking object needs to be elevated and shared across all threads/light threads that want concurrent access to the underlying resouce.
2. It is not a good idea to mix sync and async. Consider using async locking mechanisms - https://docs.rs/async-std/latest/async_std/sync/struct.MutexGuard.html

Reply:
I have addressed (2) by removing the use of the async_std crate.

Reply:
For (1), please see these examples:
https://stackoverflow.com/questions/62248219/rust-accessing-option-from-mutex
https://stackoverflow.com/questions/72018622/getting-vector-from-an-arc-mutex

You will see that the lock call returns a Result for a MutexGuard and that Result is what we unwrap.

Reply:
Regarding (2), I disagree with using async everywhere.  Allowing this is referred to as "async creep" and it is considered an anti-pattern.

https://medium.com/@davealexis/this-is-why-i-consider-the-async-await-pattern-to-be-like-a-virus-e029d95fcba1

Reply:
It's OK to have both sync and async as long as synchronous blocking is not used with async code. If you guarantee it is not the case here, then it's alright, but

On the other hand:

1. MutexGuard from async_std is similar if not the same from user perspective is the one from the sync part.
2. Other contributors/users of this code may not be as careful and run into an nondeterministic bugs when synchronously locking some async functionality.

Reply:
I agree with "the developer does not always go the extra yard to ensure that their..." put your favorite ending there. What async should do to the developer is make him/her extra alarmed and watch with increased caution for additional things that can go wrong, like locking a whole thread, for one.

I'd rather go with tasks/green threads/async all in from start because it allows to execute more concurrent things in a more frugal way in parallel (cloud distributed [IPS]aaS are that way) and don't worry about re-writing the thing later, if async is desired, but I will leave it to you whether you feel it is necessary or not for this project.

Reply:
On a second thought, this func itself is async and you are doing sync lock, which blocks a thread - that is no good. In so many places in this code async is used. You might as well go all in with async now, rather than later.

The Rust Book on Async covers whys - https://rust-lang.github.io/async-book/01_getting_started/02_why_async.html#async-vs-threads-in-rust

Reply:
Which func are you referring to?  Are you looking at the latest version of the code?

Reply:
This one, which contains this comment - https://github.com/eclipse-ibeji/ibeji/blob/ff3017cbc387c8c0021224bd5d5c19b731bc70bd/samples/mixed/provider/src/provider_impl.rs#L50

and other places - look around for async/await just in this change.

Reply:
This looks fine.

Reply:
The use of the locking is fine.

Reply:
This locks the whole thread on a function that is being awaited, non letting the context to switch. Here is a 10+ year old exlanation why this is bad https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html

Reply:
These two document provides a good discussion on when to use std::sync::Mutex and when to use tokio::sync::Mutex:

- https://tokio.rs/tokio/tutorial/shared-state
- https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html

The code uses Mutex for ensuring mutually exclusive access to data and never holds the lock over an .await point.

I switched to using parking_lot::Mutex as per the recommendations in these documents,